File: services/go-analytics/internal/services/analytics_test.go
package services

import (
	"context"
	"testing"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// TestAnalyticsEvent tests event struct creation
func TestAnalyticsEventCreation(t *testing.T) {
	projectID := primitive.NewObjectID()
	event := &AnalyticsEvent{
		ProjectID: projectID,
		Event:     "page_view",
		SessionID: "session123",
		URL:       "https://example.com/home",
		Referrer:  "https://google.com",
		UserID:    "user123",
		Data:      map[string]interface{}{"button": "signup"},
		Device:    &Device{Type: "desktop", OS: "macOS", Browser: "Chrome"},
		Geo:       &Geo{Country: "US", City: "San Francisco", Timezone: "America/Los_Angeles"},
		Timestamp: time.Now(),
	}

	if event.Event != "page_view" {
		t.Errorf("Expected event 'page_view', got '%s'", event.Event)
	}

	if event.SessionID != "session123" {
		t.Errorf("Expected sessionID 'session123', got '%s'", event.SessionID)
	}

	if event.Device.Type != "desktop" {
		t.Errorf("Expected device type 'desktop', got '%s'", event.Device.Type)
	}
}

// TestDeviceStruct tests Device struct
func TestDeviceStruct(t *testing.T) {
	tests := []struct {
		name     string
		device   Device
		expected string
	}{
		{"Desktop", Device{Type: "desktop", OS: "Windows", Browser: "Firefox"}, "desktop"},
		{"Mobile", Device{Type: "mobile", OS: "iOS", Browser: "Safari"}, "mobile"},
		{"Tablet", Device{Type: "tablet", OS: "Android", Browser: "Chrome"}, "tablet"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.device.Type != tt.expected {
				t.Errorf("Expected device type '%s', got '%s'", tt.expected, tt.device.Type)
			}
		})
	}
}

// TestGeoStruct tests Geo struct
func TestGeoStruct(t *testing.T) {
	geo := &Geo{
		Country:  "CN",
		City:     "Beijing",
		Timezone: "Asia/Shanghai",
	}

	if geo.Country != "CN" {
		t.Errorf("Expected country 'CN', got '%s'", geo.Country)
	}

	if geo.City != "Beijing" {
		t.Errorf("Expected city 'Beijing', got '%s'", geo.City)
	}
}

// TestPeriodStruct tests Period struct
func TestPeriodStruct(t *testing.T) {
	now := time.Now()
	period := Period{
		Start: now.Add(-24 * time.Hour),
		End:   now,
	}

	if period.End.Before(period.Start) {
		t.Error("End time should be after start time")
	}

	duration := period.End.Sub(period.Start)
	if duration != 24*time.Hour {
		t.Errorf("Expected duration of 24 hours, got %v", duration)
	}
}

// TestProjectStatsStruct tests ProjectStats struct
func TestProjectStatsStruct(t *testing.T) {
	stats := &ProjectStats{
		Users: UserStats{
			Total:  1000,
			New:    100,
			Active: 500,
			Change: 5.5,
		},
		PageViews: PageViewStats{
			Total:  5000,
			Change: 10.2,
		},
		Sessions: SessionStats{
			Total:       2000,
			AvgDuration: 180,
			Change:      3.1,
		},
		Conversion: ConversionStats{
			Rate:   2.5,
			Change: 0.5,
		},
		Engagement: EngagementStats{
			BounceRate:          45.0,
			AvgSessionDuration:  180,
			PageViewsPerSession: 2.5,
		},
	}

	if stats.Users.Total != 1000 {
		t.Errorf("Expected 1000 total users, got %d", stats.Users.Total)
	}

	if stats.Conversion.Rate != 2.5 {
		t.Errorf("Expected conversion rate 2.5, got %f", stats.Conversion.Rate)
	}
}

// TestTrendData tests TrendData struct
func TestTrendData(t *testing.T) {
	trends := []TrendData{
		{Date: "2024-01-01", Value: 100},
		{Date: "2024-01-02", Value: 120},
		{Date: "2024-01-03", Value: 150},
	}

	if len(trends) != 3 {
		t.Errorf("Expected 3 trend data points, got %d", len(trends))
	}

	if trends[2].Value != 150 {
		t.Errorf("Expected value 150 for last data point, got %d", trends[2].Value)
	}
}

// TestFunnelStep tests FunnelStep struct
func TestFunnelStep(t *testing.T) {
	steps := []FunnelStep{
		{Name: "visit", Count: 1000, Rate: 100, Dropoff: 0},
		{Name: "signup", Count: 200, Rate: 20, Dropoff: 80},
		{Name: "purchase", Count: 50, Rate: 25, Dropoff: 75},
	}

	if steps[0].Count != 1000 {
		t.Errorf("Expected 1000 visits, got %d", steps[0].Count)
	}

	if steps[1].Dropoff != 80 {
		t.Errorf("Expected 80%% dropoff at signup, got %f", steps[1].Dropoff)
	}
}

// TestPageView tests PageView struct
func TestPageView(t *testing.T) {
	pages := []PageView{
		{URL: "/home", Views: 500},
		{URL: "/products", Views: 300},
		{URL: "/about", Views: 100},
	}

	if pages[0].URL != "/home" {
		t.Errorf("Expected URL '/home', got '%s'", pages[0].URL)
	}

	totalViews := int32(0)
	for _, p := range pages {
		totalViews += p.Views
	}

	if totalViews != 900 {
		t.Errorf("Expected 900 total views, got %d", totalViews)
	}
}

// TestCalculateChange tests the calculateChange function
func TestCalculateChange(t *testing.T) {
	tests := []struct {
		name     string
		current  float32
		previous float32
		expected float32
	}{
		{"Positive change", 120, 100, 20},
		{"Negative change", 80, 100, -20},
		{"No change", 100, 100, 0},
		{"From zero (current > 0)", 100, 0, 100},
		{"From zero (current = 0)", 0, 0, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := calculateChange(tt.current, tt.previous)
			if result != tt.expected {
				t.Errorf("calculateChange(%f, %f) = %f; want %f", tt.current, tt.previous, result, tt.expected)
			}
		})
	}
}

// TestGenerateTrackingScript tests tracking script generation
func TestGenerateTrackingScript(t *testing.T) {
	service := &AnalyticsService{}
	projectID := "test123"

	script := service.GenerateTrackingScript(projectID)

	if script == "" {
		t.Error("Expected non-empty tracking script")
	}

	if !contains(script, projectID) {
		t.Error("Script should contain project ID")
	}

	if !contains(script, "thinkusTrack") {
		t.Error("Script should define thinkusTrack function")
	}

	if !contains(script, "page_view") {
		t.Error("Script should track page_view event")
	}
}

// TestTrackingScriptHasRequiredElements tests script elements
func TestTrackingScriptHasRequiredElements(t *testing.T) {
	service := &AnalyticsService{}
	script := service.GenerateTrackingScript("project123")

	requiredElements := []string{
		"THINKUS_PROJECT_ID",
		"THINKUS_API",
		"SESSION_ID",
		"getDeviceType",
		"navigator.sendBeacon",
		"beforeunload",
	}

	for _, elem := range requiredElements {
		if !contains(script, elem) {
			t.Errorf("Script should contain '%s'", elem)
		}
	}
}

// Helper function
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > 0 && containsHelper(s, substr))
}

func containsHelper(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

// Benchmark tests
func BenchmarkCalculateChange(b *testing.B) {
	for i := 0; i < b.N; i++ {
		calculateChange(120, 100)
	}
}

func BenchmarkGenerateTrackingScript(b *testing.B) {
	service := &AnalyticsService{}
	for i := 0; i < b.N; i++ {
		service.GenerateTrackingScript("project123")
	}
}

// Integration-ready tests (require mocked DB)
func TestNewAnalyticsService(t *testing.T) {
	// This test would require a mock MongoDB connection
	// Skip in unit test, run in integration test
	t.Skip("Requires MongoDB connection - run as integration test")
}

func TestTrack(t *testing.T) {
	// This test would require a mock MongoDB connection
	t.Skip("Requires MongoDB connection - run as integration test")
}

func TestGetStats(t *testing.T) {
	// This test would require a mock MongoDB connection
	t.Skip("Requires MongoDB connection - run as integration test")
}

func TestGetTrends(t *testing.T) {
	// This test would require a mock MongoDB connection
	t.Skip("Requires MongoDB connection - run as integration test")
}

func TestGetFunnel(t *testing.T) {
	// This test would require a mock MongoDB connection
	t.Skip("Requires MongoDB connection - run as integration test")
}

func TestGetTopPages(t *testing.T) {
	// This test would require a mock MongoDB connection
	t.Skip("Requires MongoDB connection - run as integration test")
}

// Validation tests
func TestValidProjectID(t *testing.T) {
	validIDs := []string{
		"507f1f77bcf86cd799439011",
		"507f191e810c19729de860ea",
	}

	for _, id := range validIDs {
		_, err := primitive.ObjectIDFromHex(id)
		if err != nil {
			t.Errorf("Expected valid ObjectID for '%s', got error: %v", id, err)
		}
	}

	invalidIDs := []string{
		"invalid",
		"123",
		"",
	}

	for _, id := range invalidIDs {
		_, err := primitive.ObjectIDFromHex(id)
		if err == nil {
			t.Errorf("Expected error for invalid ObjectID '%s'", id)
		}
	}
}

// Context timeout test
func TestContextTimeout(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Millisecond)
	defer cancel()

	time.Sleep(2 * time.Millisecond)

	select {
	case <-ctx.Done():
		// Expected
	default:
		t.Error("Context should have timed out")
	}
}

File: thinkus-app/tests/unit/lib/grpc/analytics.test.ts
/**
 * Tests for Analytics Service gRPC Client
 */
import { describe, it, expect, vi } from 'vitest'

describe('Analytics Service Client Types', () => {
  describe('TrackRequest', () => {
    it('should have correct structure', () => {
      const request = {
        project_id: 'project123',
        event: 'page_view',
        session_id: 'session123',
        url: 'https://example.com/page',
        referrer: 'https://google.com',
        user_id: 'user123',
        data: { button: 'signup' },
        device: { type: 'desktop', os: 'macOS', browser: 'Chrome' },
        geo: { country: 'US', city: 'San Francisco', timezone: 'America/Los_Angeles' },
      }

      expect(request.event).toBe('page_view')
      expect(request.device.type).toBe('desktop')
      expect(request.geo.country).toBe('US')
    })

    it('should support common event types', () => {
      const events = [
        'page_view',
        'session_start',
        'session_end',
        'click',
        'conversion',
        'signup',
        'purchase',
      ]

      events.forEach((event) => {
        const request = { event }
        expect(request.event).toBe(event)
      })
    })
  })

  describe('Device', () => {
    it('should support all device types', () => {
      const types = ['desktop', 'mobile', 'tablet']

      types.forEach((type) => {
        const device = { type, os: '', browser: '' }
        expect(['desktop', 'mobile', 'tablet']).toContain(device.type)
      })
    })

    it('should support common browsers', () => {
      const browsers = ['Chrome', 'Firefox', 'Safari', 'Edge', 'Opera']

      browsers.forEach((browser) => {
        const device = { type: 'desktop', os: '', browser }
        expect(device.browser).toBe(browser)
      })
    })

    it('should support common operating systems', () => {
      const systems = ['Windows', 'macOS', 'Linux', 'iOS', 'Android']

      systems.forEach((os) => {
        const device = { type: 'desktop', os, browser: '' }
        expect(device.os).toBe(os)
      })
    })
  })

  describe('Geo', () => {
    it('should have correct structure', () => {
      const geo = {
        country: 'CN',
        city: 'Beijing',
        timezone: 'Asia/Shanghai',
      }

      expect(geo.country).toBe('CN')
      expect(geo.timezone).toContain('Asia')
    })

    it('should support ISO country codes', () => {
      const countries = ['US', 'CN', 'GB', 'DE', 'JP', 'FR']

      countries.forEach((country) => {
        expect(country).toHaveLength(2)
      })
    })
  })

  describe('Period', () => {
    it('should have start and end timestamps', () => {
      const now = Date.now()
      const period = {
        start: now - 24 * 60 * 60 * 1000, // 24 hours ago
        end: now,
      }

      expect(period.end).toBeGreaterThan(period.start)
      expect(period.end - period.start).toBe(24 * 60 * 60 * 1000)
    })
  })

  describe('ProjectStats', () => {
    it('should have all required sections', () => {
      const stats = {
        users: { total: 1000, new: 100, active: 500, change: 5.5 },
        page_views: { total: 5000, change: 10.2 },
        sessions: { total: 2000, avg_duration: 180, change: 3.1 },
        conversion: { rate: 2.5, change: 0.5 },
        engagement: {
          bounce_rate: 45.0,
          avg_session_duration: 180,
          page_views_per_session: 2.5,
        },
      }

      expect(stats.users.total).toBe(1000)
      expect(stats.conversion.rate).toBe(2.5)
      expect(stats.engagement.bounce_rate).toBe(45.0)
    })
  })

  describe('TrendData', () => {
    it('should have date and value', () => {
      const trends = [
        { date: '2024-01-01', value: 100 },
        { date: '2024-01-02', value: 120 },
        { date: '2024-01-03', value: 150 },
      ]

      expect(trends).toHaveLength(3)
      expect(trends[0].date).toBe('2024-01-01')
      expect(trends[2].value).toBe(150)
    })
  })

  describe('FunnelStep', () => {
    it('should have correct structure', () => {
      const steps = [
        { name: 'visit', count: 1000, rate: 100, dropoff: 0 },
        { name: 'signup', count: 200, rate: 20, dropoff: 80 },
        { name: 'purchase', count: 50, rate: 25, dropoff: 75 },
      ]

      expect(steps[0].rate).toBe(100)
      expect(steps[1].dropoff).toBe(80)
    })

    it('should calculate dropoff correctly', () => {
      const rate = 20
      const dropoff = 100 - rate

      expect(dropoff).toBe(80)
    })
  })

  describe('StreamEvent', () => {
    it('should have correct structure', () => {
      const event = {
        type: 'code_change',
        project_id: 'project123',
        timestamp: Date.now(),
        payload: JSON.stringify({ file: 'main.ts', content: 'code' }),
      }

      expect(event.type).toBe('code_change')
      expect(event.timestamp).toBeGreaterThan(0)
    })

    it('should support all event types', () => {
      const types = [
        'code_change',
        'terminal_output',
        'agent_status',
        'progress',
        'preview_update',
        'error',
        'message',
      ]

      types.forEach((type) => {
        const event = { type }
        expect(types).toContain(event.type)
      })
    })
  })
})

describe('Analytics Service Client Functions', () => {
  // Mock gRPC
  vi.mock('@grpc/grpc-js', () => ({
    credentials: {
      createInsecure: vi.fn(() => ({})),
    },
    loadPackageDefinition: vi.fn(() => ({
      thinkus: {
        analytics: {
          AnalyticsService: vi.fn(() => ({
            Track: vi.fn(),
            TrackBatch: vi.fn(),
            GetStats: vi.fn(),
            GetTrends: vi.fn(),
            GetFunnel: vi.fn(),
            GetTopPages: vi.fn(),
            GenerateTrackingScript: vi.fn(),
            close: vi.fn(),
          })),
          RealtimeService: vi.fn(() => ({
            Subscribe: vi.fn(),
            Push: vi.fn(),
            close: vi.fn(),
          })),
        },
      },
    })),
  }))

  vi.mock('@grpc/proto-loader', () => ({
    loadSync: vi.fn(() => ({})),
  }))

  describe('track', () => {
    it('should require project_id and event', () => {
      const request = {
        project_id: 'project123',
        event: 'page_view',
      }

      expect(request.project_id).toBeDefined()
      expect(request.event).toBeDefined()
    })
  })

  describe('getStats', () => {
    it('should require project_id and period', () => {
      const request = {
        project_id: 'project123',
        period: {
          start: Date.now() - 86400000,
          end: Date.now(),
        },
      }

      expect(request.project_id).toBeDefined()
      expect(request.period.start).toBeLessThan(request.period.end)
    })
  })

  describe('getTrends', () => {
    it('should support different metrics', () => {
      const metrics = ['pageViews', 'sessions', 'users', 'conversions']

      metrics.forEach((metric) => {
        const request = { metric }
        expect(['pageViews', 'sessions', 'users', 'conversions']).toContain(request.metric)
      })
    })
  })

  describe('getFunnel', () => {
    it('should accept array of steps', () => {
      const steps = ['visit', 'signup', 'activate', 'purchase']

      expect(steps).toHaveLength(4)
      expect(steps[0]).toBe('visit')
    })
  })

  describe('getTopPages', () => {
    it('should support limit parameter', () => {
      const limits = [5, 10, 20, 50, 100]

      limits.forEach((limit) => {
        expect(limit).toBeGreaterThan(0)
      })
    })
  })

  describe('subscribe (realtime)', () => {
    it('should require project_id', () => {
      const request = { project_id: 'project123' }
      expect(request.project_id).toBeDefined()
    })
  })

  describe('push (realtime)', () => {
    it('should require project_id and type', () => {
      const request = {
        project_id: 'project123',
        type: 'code_change',
        payload: Buffer.from(JSON.stringify({ data: 'test' })),
      }

      expect(request.project_id).toBeDefined()
      expect(request.type).toBeDefined()
    })
  })
})
